iree/compiler/Dialect/Flow/IR/FlowOpUtils.h

```diff
@@ -0,0 +1,68 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "llvm/ADT/MapVector.h"
+#include "llvm/ADT/SmallVector.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/Operation.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+// Clones an operation with new result types.
+// The original operation will be erased and a new operation constructed
+// in its place.
+Operation *cloneWithNewResultTypes(Operation *op,
+                                   llvm::SmallVectorImpl<Type> &newResultTypes);
+
+// Utility class to optimize a "closure" op, which maintains a variadic
+// list of operands corresponding to entry block arguments.
+class ClosureOpDce {
+ public:
+  ClosureOpDce(Operation *closureOp, Block &entryBlock,
+               unsigned variadicOffset);
+
+  bool needsOptimization() { return needsOperandElision || needsResultElision; }
+
+  // Whether the operation needs to be replaced.
+  bool needsNewOperation() { return needsResultElision; }
+
+  // Performs the optimization. If the optional eraseOriginal=false and
```

Comment:
eraseOriginal is unfortunate - is that because you want to pass in an OpBuilder instead of a PatternRewriter?
(really, it feels like Builder should just have all these utilities so it doesn't matter...)

Reply:
Added a TODO to fix upstream. I'm not going to get creative locally since I know this class hierarchy is getting fixed and trying to work around locally will just make it harder to integrate the future cleanup.